Create release workflow

Felixoid opened the pull request:
Changelog category (leave one):
- Build/Testing/Packaging Improvement


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add action for published releases 


Detailed description / Documentation draft:
When releases `v\d\d[.]\d+[.]\d+[.]\d+-(testing|prestable|stable|lts)` are published, the workflow will:
- Download binary packages from S3
- Upload them to artifactory
- Push them to release assets

Felixoid commented:
The workflow to check it works

alesapin commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,55 @@
+name: DocsReleaseChecks
```
Let's rename file too -- `docs_release.yml`

alesapin commented on .github/workflows/release.yml:
```diff
@@ -1,55 +1,32 @@
-      - 'docker/**'
+  release:
+    types:
+    - published
```
Works for both `pre-release` and `release`?

alesapin commented on .github/workflows/release.yml:
```diff
@@ -1,55 +1,32 @@
+        TEMP_PATH: ${{runner.temp}}/release_packages
+      run: |
+        rm -rf "$TEMP_PATH" && mkdir -p "$TEMP_PATH"
+        python3 ./tests/ci/push_to_artifactory.py --release "${{ github.ref }}" \
```
Maybe let's make a repo copy? What if your script changes something in repo directory?

Felixoid replied:
done

alesapin reviewed (commented):
But we still need to do all other release-related staff:
1. We have to manually update several files in repository. Look at the recent commit from the old release script for tweak releases: https://github.com/ClickHouse/ClickHouse/pull/30978/commits/9b13386cd4e53886785856a96ff189063590e7be and https://github.com/ClickHouse/ClickHouse/pull/30978/commits/1a254b97e2b56275fe89416febc07a946007f13a.
2. For new patch release (21.11 -> 21.12) we also need to create new PR like https://github.com/ClickHouse/ClickHouse/pull/32419 and change something in master https://github.com/ClickHouse/ClickHouse/commit/f2ccddb7be65179154d9b80a962c060828934529.
3. Push new docker (`clickhouse/clickhouse-server` and `clickhouse/clickhouse-client`) images to dockerhub.
4. Some docs how to create release.

Felixoid committed:
Rename release action to docs to reflect the content

Felixoid committed:
Improve push_to_artifactory

- Add tgz archives
- Add --all argument for all packages download
- Add -n to skip pushing to artifactory
- Add help for artifactory-url
- Make 3.8 compatible
- Add cleanup `refs/tags/` from --release argument

Felixoid committed:
Create release workflow

Felixoid committed:
Fix some linter issues

Felixoid committed:
Add option to skip redownload, push tgz

Felixoid committed:
Improve typing and arguments

Felixoid commented:
The arm64 packages will be addressed in #33664

alesapin reviewed (approved):
But we will need to update docs_release workflow id in rerun lambda.

Felixoid commented:
For history, now DocksReleaseCheck is:

```
    {
      "id": 14738810,
      "node_id": "W_kwDOA5dJV84A4OV6",
      "name": "DocsReleaseChecks",
      "path": ".github/workflows/release.yml",
      "state": "active",
      "created_at": "2021-10-30T08:21:55.000Z",
      "updated_at": "2022-01-17T23:04:40.000Z",
      "url": "https://api.github.com/repos/ClickHouse/ClickHouse/actions/workflows/14738810",
      "html_url": "https://github.com/ClickHouse/ClickHouse/blob/master/.github/workflows/release.yml",
      "badge_url": "https://github.com/ClickHouse/ClickHouse/workflows/DocsReleaseChecks/badge.svg"
    },
```
